shared/health/src/main/java/io/pravega/shared/health/impl/HealthContributorImpl.java

```diff
@@ -0,0 +1,162 @@
+ * that depend on (or are composed of) other {@link HealthContributor} objects. To do this we need to due two things:
+ * reducing/reconciling the (potentially) many different {@link Status} states and a way to persist these relationships.
+ *
+ * {@link HealthContributorImpl} objects are used *strictly* for grouping, meaning that they do not define their own
+ * health checking logic and therefore does not export any details ({@link Health#getDetails}).
+ */
+@Slf4j
+@ThreadSafe
+public abstract class HealthContributorImpl implements HealthContributor {
+
+    /**
+     * The {@link StatusAggregator} used to perform the aggregation of all the {@link HealthContributor} dependencies.
+     */
+    @Getter
+    @NonNull
+    private final StatusAggregator aggregator;
+
+    /**
+     * The identifier for this {@link HealthContributor}.
+     */
+    @Getter
+    @NonNull
+    private final String name;
+
+    /**
+     * The most recent reported status, either by the {@link #doHealthCheck(Health.HealthBuilder)} or an explicit mutation.
+     * Useful when there is a sudden failure and can provide a distinction between an expected closure/shutdown
+     * and an abrupt 'failing' closure.
+     */
+    @Getter
+    @Setter
+    private Status status;
+    /**
+     * Flag indicating this contributor and all its children are no longer accessible.
+     */
+    private final AtomicBoolean closed = new AtomicBoolean();
+
+    /**
+     * Holds the references to all child {@link HealthContributor}.
+     */
+    private final Map<String, HealthContributor> contributors = new ConcurrentHashMap<>();
+
+    public HealthContributorImpl(@NonNull String name) {
+        this(name, StatusAggregator.UNANIMOUS);
+    }
+
+    public HealthContributorImpl(@NonNull String name, @NonNull StatusAggregator aggregator) {
+        this.name = name;
+        this.aggregator = aggregator;
+    }
+
+    /**
+     * Recursively build (in post-order fashion) the {@link Health} result for a given {@link HealthContributor}.
+     *
+     * @return The {@link Health} result of the {@link HealthContributor}.
+     */
+    @Override
+    synchronized final public Health getHealthSnapshot() {
+        Exceptions.checkNotClosed(isClosed(), this);
+
+        Health.HealthBuilder builder = Health.builder().name(getName());
+        Collection<Status> statuses = new ArrayList<>();
+        Map<String, Health> children = new HashMap<>();
+
+        for (val  contributor : contributors.entrySet()) {
+            if (!contributor.getValue().isClosed()) {
+                Health health =  contributor.getValue().getHealthSnapshot();
+                children.put(contributor.getKey(), health);
+                statuses.add(health.getStatus());
+            } else {
+                contributors.remove(name);
+            }
+        }
+
+        Status status = Status.DOWN;
+        // Perform own health check logic.
+        try {
+            status = doHealthCheck(builder);
```

Comment:
`getHealthSnapshot()` returns a `Health` object for a component and all its subcomponents. Since doHealthCheck() has actual health check implementation for 1 component that could be subcomponent of another one or a top level component I would expect this to return a Health object going by the composite design pattern.

Reply:
`doHealthCheck` is responsible for a) Determining the status of some arbitrary component/class and b) declaring the details you wish to export for said component/class. The important part is this is done _without_ consideration of its children components, so it cannot build a complete `Health` object. `getHealthSnapshot` reconciles the children and the parent to build the `Health` object. I don't see any violation of the 'composite design pattern'.